File: scrips/topic_bridge.py
import rospy
from std_msgs.msg import String
from lora import LoRa

# Initialize ROS2 Node
rospy.init_node('ros2_lora_topic_bridge')

# Set LoRa parameters
frequency = 915.0
spreading_factor = 7
coding_rate = 5
bw = 125000
sync_word = 0x34
tx_power = 17
gpio_pins = {'dio_0': 17, 'reset': 11, 'spi': {'mosi': 10, 'miso': 9, 'sclk': 11, 'cs': 8}}

# Initialize LoRa
lora = LoRa(frequency=frequency, spreading_factor=spreading_factor, coding_rate=coding_rate,
            bw=bw, sync_word=sync_word, tx_power=tx_power, gpio_pins=gpio_pins)

# ROS2 Publisher Callback
def ros2_topic_callback(msg):
    # Convert ROS2 message to string
    data = msg.data
    # Send data over LoRa
    lora.send(data)

# ROS2 Subscriber
ros2_subscriber = rospy.Subscriber('/ros2_topic', String, ros2_topic_callback)

# LoRa Receive Callback
def lora_receive_callback(data):
    # Convert LoRa data to string
    data = data.decode('utf-8')
    # Publish data on ROS2 topic
    ros2_publisher.publish(String(data=data))

# LoRa Receiver
lora.set_receive_callback(lora_receive_callback)

# ROS2 Publisher
ros2_publisher = rospy.Publisher('/lora_topic', String, queue_size=10)

# ROS2 Spin
rospy.spin()

File: src/topic_bridge.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "lora.h" // Include the LoRa library

class Ros2LoRaTopicBridge : public rclcpp::Node
{
public:
  Ros2LoRaTopicBridge() : Node("ros2_lora_topic_bridge")
  {
    // Set LoRa parameters
    double frequency = 915.0;
    int spreading_factor = 7;
    int coding_rate = 5;
    int bw = 125000;
    int sync_word = 0x34;
    int tx_power = 17;
    // Initialize LoRa
    lora_ = new LoRa(frequency, spreading_factor, coding_rate, bw, sync_word, tx_power);

    // ROS2 Publisher
    ros2_publisher_ = this->create_publisher<std_msgs::msg::String>("/lora_topic", 10);

    // ROS2 Subscriber
    ros2_subscriber_ = this->create_subscription<std_msgs::msg::String>(
        "/ros2_topic", 10,
        std::bind(&Ros2LoRaTopicBridge::ros2TopicCallback, this, std::placeholders::_1));

    // LoRa Receive Callback
    lora_->setReceiveCallback(
        std::bind(&Ros2LoRaTopicBridge::loraReceiveCallback, this, std::placeholders::_1));
  }

private:
  void ros2TopicCallback(const std_msgs::msg::String::SharedPtr msg)
  {
    // Convert ROS2 message to string
    std::string data = msg->data;
    // Send data over LoRa
    lora_->send(data);
  }

  void loraReceiveCallback(const std::vector<uint8_t> &data)
  {
    // Convert LoRa data to string
    std::string received_data(data.begin(), data.end());
    // Publish data on ROS2 topic
    auto ros2_msg = std::make_shared<std_msgs::msg::String>();
    ros2_msg->data = received_data;
    ros2_publisher_->publish(ros2_msg);
  }

  LoRa *lora_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr ros2_publisher_;
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr ros2_subscriber_;
};

int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<Ros2LoRaTopicBridge>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
